[3.1] Add IntegrationTestCase::cookieEncrypted().

waterada opened the pull request:
Add `IntegrationTestCase::cookieEncrypted()` to encrypt with the CookieComponent.
See: issue #7862 
See also: the previous pull request #7873 (will be closed soon)
- This is the re-pull-request which is based on master on [@lorenzo's advice](https://github.com/cakephp/cakephp/pull/7873#issuecomment-165981425).
- Extracting the crypt logic from `CookieComponent` to the trait on [@markstory's advice](https://github.com/cakephp/cakephp/pull/7873#discussion_r48092730).
- Add the new method `cookieEncrypted()` separately from `cookie()`  on [@markstory's advice](https://github.com/cakephp/cakephp/pull/7873#discussion_r48094276).
-  Add the test to make sure a controller action can decode the cookie on [@markstory's advice](https://github.com/cakephp/cakephp/pull/7873#discussion_r48092742).
- Add the forth argument to specify the encryption key to `cookieEncrypted()` on [@jeremyharris' advice](https://github.com/cakephp/cakephp/pull/7873#issuecomment-165992475).

waterada committed:
Add IntegrationTestCase::cookieEncrypted().

waterada commented:
@jeremyharris The `CookieComponent` encrypt values by default, so I think most users use it without the conscious of the configuration of `CookieComponent`.
Therefore, I think it is nicer to use it without `['encryption' => ***]`.
However, the case that you need to pass the encryption key can be possible. So, I have added the forth argument to do this. How about this?

waterada committed:
Fix: InvalidArgumentException: Invalid key for encrypt(), key must be at least 256 bits (32 bytes) long.

markstory commented on src/TestSuite/IntegrationTestCase.php:
```diff
@@ -219,6 +229,43 @@ public function cookie($name, $value)
+        if (isset($key)) {
+            return $key;
+        }
+        $key = Security::salt('foo!foo!foo!foo!foo!foo!foo!foo!');
```
Why not a random string?

waterada replied:
If beginners forget `Security::salt('...')` before `cookieEncrypted()`, they will see `InvalidArgumentException: Invalid key for encrypt(), key must be at least 256 bits (32 bytes) long.` I think it is a little difficult for them to resolve it by themselves. If there is this default salt, they can `cookieEncrypted()` even if they forget to set the salt. It is the reason why I introduced this default salt. The random string is no meaning. I just wanted the 32 bytes string. How do you feel this? Should I remove this?

markstory replied:
In some ways an exception is more useful as the code will not function outside of the test suite. Providing a default is easier for the user, but could mask configuration issues.

gr-terada replied:
I see. I will remove it.

markstory commented on src/TestSuite/IntegrationTestCase.php:
```diff
@@ -219,6 +229,43 @@ public function cookie($name, $value)
+    /**
+     * Sets a encrypted request cookie for future requests.
+     *
+     * The difference from cookie() is this encrypts the coolie
```
coolie should be cookie.

waterada replied:
I will fix it!

markstory commented on tests/TestCase/Controller/CookieComponentControllerTest.php:
```diff
@@ -0,0 +1,124 @@
+        $this->get('/cookie_component_test/view/');
+        $this->assertStringStartsWith('Q2FrZQ==.', $this->viewVariable('ValueFromRequest'), 'Encrypted');
+        $this->assertEquals(['Value1 of Cookie', 'Value2 of Cookie'], $this->viewVariable('ValueFromCookieComponent'),
+            'Decrypted');
```
The PHPCS build will want each argument on separate lines.

markstory commented on tests/TestCase/Controller/CookieComponentControllerTest.php:
```diff
@@ -0,0 +1,124 @@
+        return [
+            ['foo!foo!foo!foo!foo!foo!foo!foo!'],
+            ['bar!bar!bar!bar!bar!bar!bar!bar!'],
+        ];
```
What is different about these two cases?

waterada replied:
I thought this is to avoid the hard coding salt. But now, I realize it certainly has no meaning. I will remove this dataProvider. Can I use `foo!foo!foo!foo!foo!foo!foo!foo!` as the salt on test case?

markstory commented on tests/test_app/TestApp/Controller/CookieComponentTestController.php:
```diff
@@ -0,0 +1,46 @@
+     */
+    public $components = [
+        'Cookie' => [
+            //'encryption' => 'aes', // default
```
Just delete commented out code.

waterada replied:
OK. I will remove it.

waterada committed:
Fix typo

waterada committed:
Add the line break at the end of file.

waterada committed:
Remove the commented out code.

waterada committed:
Add a salt initialization.

jeremyharris commented:
I think that will work since I can't see why we'd need the other options during tests. I was aiming for the signature to be a bit more future proof or even accept a builder class, but if everyone's okay with more args, I am :)

waterada committed:
Fix: Missing function doc comment
Fix: Opening brace should be on a new line

waterada committed:
Fix: Use classes must be in alphabetical order.

waterada committed:
Fix: Opening brace should be on a new line.

waterada committed:
Fix: Opening parenthesis of a multi-line function call must be the last content on the line
Fix: Only one argument is allowed per line in a multi-line function call
Fix: Multi-line function call not indented correctly; expected 8 spaces but found 12
Fix: Closing parenthesis of a multi-line function call must be on a line by itself

waterada committed:
Remove useless cases.

waterada committed:
Remove the over-parented default salt. It might mask configuration issues.

markstory commented on tests/TestCase/Controller/CookieComponentControllerTest.php:
```diff
@@ -0,0 +1,112 @@
+/**
+ * CookieComponentControllerTest class
+ */
+class CookieComponentControllerTest extends IntegrationTestCase
```
This test case should probably be a sibling of `tests/TestCase/TestSuite/`. It looks like an integration test for the integration test case :smile:

gr-terada replied:
OK! I will move it to `tests/TestCase/TestSuite/`!

waterada committed:
Move/rename to where/what should be so that the meaning becomes clearer.

markstory commented:
Looks good to me.
